### ISSUE TITLE
eval_shape appears to hang when tracing a large random sample

### ISSUE THREAD

**Issue by davmre** [google/jax#7011]
Calling

```python
import jax
from jax import random
jax.eval_shape(lambda: random.normal(key=random.PRNGKey(0), shape=[int(1e11)]))
```

returns `ShapeDtypeStruct(shape=(100000000000,), dtype=float32)`. I'd expect that calling

```python
jax.eval_shape(lambda: random.normal(key=random.PRNGKey(0), shape=[int(1e11), int(1e11)]))
```

would similarly return `ShapeDtypeStruct(shape=(100000000000,100000000000), dtype=float32)` (of course, if we *actually* ran the function we'd get an OOM error). 

But instead, this call seems to end up in an ~infinite loop (at least, it hangs for > 2 minutes). Reproduction here:
https://colab.research.google.com/drive/14IZF3WUTaq976jaUC5tehKTPwt3YhNTe?usp=sharing

Is this expected?

**Comment by jakevdp** [comment]
It looks like this comes from the implementation of `_random_bits`, specifically this part: https://github.com/google/jax/blob/1d0c461ea67db683206d2643a5b377f2f8ff2e84/jax/_src/random.py#L306-L312

If the number of generated numbers is larger than the largest uint32, the implementation stacks multiple calls to the underlying random number generator in a Python loop. Since these calls are required for abstract evaluation, they are generated even for `jax.eval_shape`. And if the size is large enough, the abstract evaluation graph will become quite large and slow to evaluate.

You can see this if you wish by calling `make_jaxpr()` for your input.

We could maybe think about fixing this, but it does seem like a pretty rare corner case. In what context did you encounter the problem?

**Comment by jakevdp** [comment]
The fix did not turn out to be so bad: see #7012

**Comment by davmre** [comment]
Thanks for looking into this! The context was that we (TFP) are using `eval_shape` and its TF equivalent to inspect properties of joint distribution models, and to test this we wrote code like the above example that's intended to work under `eval_shape` but crash if the ops are actually executed. So it was surprising when the code also didn't work under `eval_shape`. :-) Not mission critical since it's just for a test, but it'll be nice to have this working!
